### PR TITLE
rustdoc-json: Refractor and document Id's

### LINKED ISSUE #133780: rustdoc-json: Document why Id can't just be DefId
> NIT: This comment could be more useful. The problem it's actually trying to solve is that a single `pub use` with 1 defid can expand to multiple rustdoc-json items it the name exists in multiple namespaces: https://godbolt.org/z/jEnGxoWx6.
>
> At some point in the future, we should maybe change how we do this on the JSON side, but not in this PR.

_Originally posted by @aDotInTheVoid in https://github.com/rust-lang/rust/pull/130078#discussion_r1796986306_

Rustdoc-json's Id handling is a bit subtle. There should be a nice doc-comment to explain it.

[It came up on zulip here](https://rust-lang.zulipchat.com/#narrow/channel/266220-t-rustdoc/topic/linking.20to.20assoc.20fn.20of.20dyn.20trait/near/485729187)

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
Closes #133780

While working on documenting Id's, I realized alot of the way they were generated was weird and unnecessary. Eg:

1. The fully uninterned id type was `(FullItemId, Option<FullItemId>)`, meaning it wasn't actually full!
2. None of the extra fields in `Option<FullItemId>` would ever be used
3. `imported_item_id` was a `rustdoc_json_types::Id` instead of a simpler `DefId`.

I believe the new implementation still covers all the same cases, but in a more principled way (and explaining why each piece is needed).

This was written to be reviewed commit-by-commit, but it might be easier to review all at once if you're not interested in tracking how the original code became the final code.

cc @its-the-shrimp

r? @fmease

### CODE REVIEW

File: src/librustdoc/json/ids.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    /// Needed for `rustc_doc_primitive` modules.
+    ///
+    /// For these, 1 [`DefId`] is used for both the primitive and the fake-module
+    /// that holds it's docs.
+    ///
+    /// N.B. This only matters when documenting the standard library with
+    /// `--document-private-items`. Maybe we should delete that module, and
+    /// remove this.
+    name: Option<Symbol>,
+}
```
Comment (its-the-shrimp) [inline src/librustdoc/json/ids.rs:66]: could this struct be an enum? I'd expect the fields `extra_id` & `name` to never be `Some` simultaneously

Comment (its-the-shrimp) [reply src/librustdoc/json/ids.rs:66]: Also, do we need to store the name? if this only exists to generate 2 rustdoc IDs for what is 1 ID in rustc, this field could be a boolean, say, `is_primitive`, or just an enum variant

Comment (aDotInTheVoid) [reply src/librustdoc/json/ids.rs:66]: I guess we could do that, and end up with something like `clean::ItemId` again.

Now I'm almost curious if we could use `clean::ItemId` in place of `json::ids::FullItemId`.

Comment (its-the-shrimp) [reply src/librustdoc/json/ids.rs:66]: Probably not, `FullItemId` would also need a `Primitive` variant then

Comment (aDotInTheVoid) [reply src/librustdoc/json/ids.rs:66]: I think it's best to leave that to a followup then. The structure's already sufficiently different.

File: src/librustdoc/json/ids.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
+//! Id handling for rustdoc-json.
+//!
+//! Manages the creation of [`rustdoc_json_types::Id`] and the
+//! fact that these don't corespond exactly to [`DefId`], because
+//! [`rustdoc_json_types::Item`] doesn't corespond exactly to what
+//! other phasises thing of as an "item".
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:6]:
```suggestion
//! Manages the creation of [`rustdoc_json_types::Id`] and the
//! fact that these don't correspond exactly to [`DefId`], because
//! [`rustdoc_json_types::Item`] doesn't correspond exactly to what
//! other phases think of as an "item".
```

File: src/librustdoc/json/ids.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+/// An uninterned id.
+///
+/// One of these coresponds to every:
+/// 1. [`rustdoc_json_types::Item`].
+/// 2. [`rustdoc_json_types::Id`] transitivly (as each `Item` has an `Id`).
+///
+/// It's *broadly* equivalent to a [`DefId`], but needs slightly more information
+/// to fully disambiguate items, because sometimes we choose to split a single HIR
+/// item into multiple JSON items, or have items with no coresponding HIR item.
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:28]:
```suggestion
/// An uninterned id.
///
/// One of these corresponds to every:
/// 1. [`rustdoc_json_types::Item`].
/// 2. [`rustdoc_json_types::Id`] transitively (as each `Item` has an `Id`).
///
/// It's *broadly* equivalent to a [`DefId`], but needs slightly more information
/// to fully disambiguate items, because sometimes we choose to split a single HIR
/// item into multiple JSON items, or have items with no corresponding HIR item.
```

File: src/librustdoc/json/ids.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+use crate::clean;
+
+pub(super) type IdInterner = FxHashMap<FullItemId, types::Id>;
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+/// An uninterned id.
+///
+/// One of these coresponds to every:
+/// 1. [`rustdoc_json_types::Item`].
+/// 2. [`rustdoc_json_types::Id`] transitivly (as each `Item` has an `Id`).
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:24]: I would rephrase it like "Each one corresponds to an:" because "one […] corresp. to every" sounds like a one-to-many relation to me (but I'm L2) while this is is a one-to-one-to-one relation, right?

File: src/librustdoc/json/ids.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+///
+/// It's *broadly* equivalent to a [`DefId`], but needs slightly more information
+/// to fully disambiguate items, because sometimes we choose to split a single HIR
+/// item into multiple JSON items, or have items with no coresponding HIR item.
+pub(super) struct FullItemId {
+    /// The "main" id of the item.
+    ///
+    /// In most cases this unequely identifies the item, other fields are just
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:32]:
```suggestion
    /// In most cases this uniquely identifies the item, the other fields are just
```

File: src/librustdoc/json/ids.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    /// used for edge-cases.
+    def_id: DefId,
+
+    /// An extra [`DefId`], which we need for:
+    ///
+    /// 1. Auto-trait impls synthesized by rustdoc.
+    /// 2. Blanket impls synthesized by rustdoc.
+    /// 3. Splitting of reexports of multiple items.
+    ///
+    ///    Eg:
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:42]:
```suggestion
    ///    E.g:
```

File: src/librustdoc/json/ids.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    ///     mod module {
+    ///         pub struct Foo {} // Exists in type namespace
+    ///         pub fn Foo(){} // Exists in value namespace
+    ///     }
+    ///
+    ///     pub use module::Foo; // Imports both items
+    ///     ```
+    ///
+    ///     In HIR, the `pub use` is just 1 item, but in rustdoc-json it's 2, so
+    ///     we need to disambiguate.
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:54]: These lines are intended by one extra space

File: src/librustdoc/json/ids.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    extra_id: Option<DefId>,
+
+    /// Needed for `rustc_doc_primitive` modules.
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:57]:
```suggestion
    /// Needed for `#[rustc_doc_primitive]` modules.
```

maybe?

File: src/librustdoc/json/ids.rs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    ///
+    /// For these, 1 [`DefId`] is used for both the primitive and the fake-module
+    /// that holds it's docs.
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:60]:
```suggestion
    /// that holds its docs.
```

File: src/librustdoc/json/ids.rs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        &self,
+        item_id: clean::ItemId,
+        name: Option<Symbol>,
+        imported_id: Option<DefId>,
+    ) -> types::Id {
+        let (def_id, mut extra_id) = match item_id {
+            clean::ItemId::DefId(did) => (did, None),
+            clean::ItemId::Blanket { for_, impl_id } => (for_, Some(impl_id)),
+            clean::ItemId::Auto { for_, trait_ } => (for_, Some(trait_)),
+        };
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:83]: What if you wrote it like this:

```suggestion
        let (def_id, extra_id) = match item_id {
            clean::ItemId::DefId(did) => (did, imported_id),
            clean::ItemId::Blanket { for_, impl_id } => (for_, Some(impl_id)),
            clean::ItemId::Auto { for_, trait_ } => (for_, Some(trait_)),
        };
```

File: src/librustdoc/json/ids.rs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+                } else {
+                    self.tcx.opt_item_name(def_id)
+                }
+            }
+        };
+
+        if let Some(imported_id) = imported_id {
+            assert_eq!(extra_id, None, "On an import item, so won't have extra from impl");
+            extra_id = Some(imported_id);
+        }
```
Comment (fmease) [inline src/librustdoc/json/ids.rs:107]: ...then you can remove this entirely, right?

```suggestion
```

**Review by fmease — APPROVED** [review]
I'm very sorry that this took so long, this shouldn't happen in the future! The changes look good.

r=me with nitpicks addressed and squashed into ~1 commit (the `SingleItemId` excursion needn't be in the final history :)).
